rama-http/src/headers/x_robots_tag_components/robots_tag_components/builder.rs

```diff
@@ -0,0 +1,136 @@
+use crate::headers::util::value_string::HeaderValueString;
+use crate::headers::x_robots_tag_components::custom_rule::CustomRule;
+use crate::headers::x_robots_tag_components::max_image_preview_setting::MaxImagePreviewSetting;
+use crate::headers::x_robots_tag_components::robots_tag::RobotsTag;
+use crate::headers::x_robots_tag_components::valid_date::ValidDate;
+use rama_core::error::OpaqueError;
+
+macro_rules! builder_field {
+    ($field:ident, $type:ty) => {
+        paste::paste! {
+            pub(in crate::headers::x_robots_tag_components) fn [<$field>](mut self, [<$field>]: $type) -> Self {
+                self.content.[<set_ $field>]([<$field>]);
+                self.valid = true;
+                self
+            }
+
+            pub(in crate::headers::x_robots_tag_components) fn [<set_ $field>](&mut self, [<$field>]: $type) -> &mut Self {
+                self.content.[<set_ $field>]([<$field>]);
+                self.valid = true;
+                self
+            }
+        }
+    };
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub(in crate::headers::x_robots_tag_components) struct Builder<T = ()> {
+    content: T,
+    valid: bool,
+}
+
+impl Builder<()> {
+    pub(in crate::headers::x_robots_tag_components) fn new() -> Self {
+        Builder {
+            content: (),
+            valid: false,
+        }
+    }
+
+    pub(in crate::headers::x_robots_tag_components) fn bot_name(
+        &self,
+        bot_name: Option<HeaderValueString>,
+    ) -> Builder<RobotsTag> {
+        Builder {
+            content: RobotsTag::new_with_bot_name(bot_name),
+            valid: false,
+        }
+    }
+}
+
+impl Builder<RobotsTag> {
```

Comment:
The beautiful thing about builder pattern in Rust is that you can make a state machine with it such that you can never ever require building into a Result type for simple cases like this, as you can prevent that an invalid usage would not compile in the first place.

E.g. someting like this:

```
pub struct NoTag {
     name: Option<HeaderValueString>,
}

impl builder<NoTag> {
   // methods to set name, just sets name (as `&mut Self` and `mut Self`), these keep the builder type unchanged

   // also methods to set one of the possible tags -> these change it to `builder<RobotsTag>`
}

impl builder<RobotsTag> {
    // methods for robot name as well as tags
    // and also allow build as at this stage we can only ever get if a tag was set :)
}
```